src/internals/ir/callGraph.ts

```diff
@@ -0,0 +1,220 @@
+import { unreachable } from "../util";
+import { TactASTStore } from "./astStore";
+import { IdxGenerator } from "./indices";
+import {
+  AstFunctionDef,
+  AstReceiver,
+  AstContractInit,
+  AstStatement,
+  AstStaticCall,
+  AstMethodCall,
+  AstExpression,
+  AstStructFieldInitializer,
+} from "@tact-lang/compiler/dist/grammar/ast";
+
+class CGEdge {
+  public idx: number;
+  public src: number;
+  public dst: number;
+
+  constructor(src: number, dst: number) {
+    this.idx = IdxGenerator.next("cg_edge");
+    this.src = src;
+    this.dst = dst;
+  }
+}
+
+class CGNode {
+  public idx: number;
+  constructor(
+    public astId: number,
+    public name: string,
+  ) {
+    this.idx = IdxGenerator.next("cg_node");
+  }
+}
+
+export class CallGraph {
+  public nodes: CGNode[] = [];
+  public edges: CGEdge[] = [];
+  private nodeMap: Map<number, CGNode> = new Map();
+  private edgeMap: Map<number, Set<number>> = new Map();
+  private astStore!: TactASTStore;
+
+  constructor() {}
+  build(astStore: TactASTStore): CallGraph {
+    this.astStore = astStore;
+    this.addFunctionsToNodes();
+    this.analyzeFunctionCalls();
+    return this;
+  }
+
+  private addFunctionsToNodes() {
+    for (const func of this.astStore.getFunctions()) {
+      const funcName = this.getFunctionName(func);
+      if (funcName) {
+        const node = new CGNode(func.id, funcName);
+        this.nodes.push(node);
+        this.nodeMap.set(func.id, node);
+      }
+    }
+  }
+
+  private analyzeFunctionCalls() {
+    for (const func of this.astStore.getFunctions()) {
+      const funcName = this.getFunctionName(func);
+      if (funcName) {
+        this.processStatements(func.statements, func.id);
+      }
+    }
+  }
+
+  private getFunctionName(
+    func: AstFunctionDef | AstReceiver | AstContractInit,
+  ): string | undefined {
+    switch (func.kind) {
+      case "function_def":
+        return func.name.text;
+      case "contract_init":
```

Comment:
What if it contains multiple contracts and receivers?

Reply:
Add some unique identifier to each contract or receiver.